File: 3501-delete-nodes-from-linked-list-present-in-array/3501-delete-nodes-from-linked-list-present-in-array.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def modifiedList(self, nums: List[int], head: Optional[ListNode]) -> Optional[ListNode]:
        ret = None
        end = None
        cur = head
        seen = []
        notSeen = []
        numsDict = {}
        for num in nums:
            numsDict[num] = True

        def findVal(val):
            # if val in seen:
            #     return True
            # if val in notSeen:
            #     return False
            # if val in nums:
            #     seen.append(val)
            #     return True
            # notSeen.append(val)
            # return False
            if val in numsDict:
                return True
            else:
                return False

        while cur.next:
            if not findVal(cur.val):
                if not ret:
                    ret = cur
                    end = cur
                else:
                    end.next = cur
                    end = end.next
            cur = cur.next
        if findVal(cur.val):
            end.next = None
        else:
            if not ret:
                ret = cur
                end = cur
            else:
                end.next = cur
        return ret



        
                
                


        

File: 3501-delete-nodes-from-linked-list-present-in-array/3501-delete-nodes-from-linked-list-present-in-array.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
 #include <unordered_set>
 using namespace std;

    //  def modifiedList(self, nums: List[int], head: Optional[ListNode]) -> Optional[ListNode]:
    //     n = set(nums)
    //     dummy = ListNode(0)
    //     dummy.next = head
    //     curr = dummy
    //     while curr.next:
    //         if (curr.next.val in n):
    //             curr.next = curr.next.next
    //         else:
    //             curr = curr.next
    //     return dummy.next
class Solution {
public:
    ListNode* modifiedList(vector<int>& nums, ListNode* head) {
        unordered_set<int> const  n(begin(nums), end(nums));
        ListNode dummy(0, head);
        ListNode* curr = &dummy;

        while (curr->next) {
            if (n.count(curr->next->val)) {
                curr->next = curr->next->next;
            } else {
                curr = curr->next;
            }
        }
        return dummy.next; 
    }
};